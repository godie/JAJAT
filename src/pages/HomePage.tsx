// src/pages/HomePage.tsx
import React, { useState, useEffect, useCallback } from 'react';
import Header from '../components/Header';
import ApplicationTable from '../components/ApplicationTable';
import TimelineView from '../components/TimelineView';
import ViewSwitcher, { type ViewType } from '../components/ViewSwitcher';
import { AlertProvider } from '../components/AlertProvider';
import { getApplications, saveApplications, generateId, type JobApplication } from '../utils/localStorage';
import AddJobForm from '../components/AddJobComponent';

const initialColumns = [
    'Position', 'Company', 'Salary', 'Status', 'Application Date', 
    'Interview Date', 'Platform', 'Contact Name', 'Follow-up Date', 'Notes', 'Link'
];

// Componente Placeholder para la sección de métricas
const MetricsSummary: React.FC<{ applications: JobApplication[] }> = ({ applications }) => {
  const totalApplications = applications.length;
  const interviews = applications.filter(a => a.interviewDate);
  const offers = applications.filter(a => a.status === 'Offer');

  const metrics = [
        { label: 'Applications', value: totalApplications, color: 'border-blue-500' },
        { label: 'Interviews', value: interviews.length, color: 'border-yellow-500' },
        { label: 'Offers', value: offers.length, color: 'border-green-500' },
    ];

  return (
    <section className="grid grid-cols-1 sm:grid-cols-2 lg:grid-cols-3 gap-6 my-8" data-testid="metrics-summary">
      {metrics.map((metric) => (
        <div 
          key={metric.label} 
          className={`bg-white p-6 rounded-xl shadow-lg border-l-4 ${metric.color} transition duration-300 hover:shadow-xl`}
        >
          <p className="text-sm font-medium text-gray-500">{metric.label}</p>
          <p className="mt-1 text-4xl font-extrabold text-gray-900">{metric.value}</p>
        </div>
      ))}
    </section>
  );
};

const HomePage: React.FC = () => {
  const [applications, setApplications] = useState<JobApplication[]>([]);
  const [currentApplication, setCurrentApplication] = useState<JobApplication | null>(null);
  const [currentView, setCurrentView] = useState<ViewType>('table');
  const isFormOpen = currentApplication !== null;

  useEffect(() => {
    setApplications(getApplications());
  }, []);

  const handleSaveEntry = useCallback((entryData: Omit<JobApplication, 'id'> | JobApplication) => {
    let newApplications:JobApplication[];

    if ('id' in entryData) {
      newApplications = applications.map(app => 
            app.id === entryData.id ? (entryData as JobApplication) : app
        );
    }
    else {
       const newEntry: JobApplication = {
      ...entryData,
      id: generateId(), // Generar ID único
      // Initialize timeline if not present
      timeline: 'timeline' in entryData ? entryData.timeline : [],
    } as JobApplication;
    newApplications = [...applications, newEntry];
    }
   

    setApplications(newApplications);
    saveApplications(newApplications);
    setCurrentApplication(null);
  }, [applications]);

  const handleDeleteEntry = useCallback((id: string) => {
    const newApplications = applications.filter(app => app.id !== id);
    setApplications(newApplications);
    saveApplications(newApplications);
  }, [applications]);

  const handleEdit = (appToEdit: JobApplication | null) => {
    setCurrentApplication(appToEdit);
  }

  const handleCreateNew = () => {
    setCurrentApplication({} as JobApplication); // Usar un objeto vacío (no nulo) para CREAR
  }
  const handleCancel = () => {
    setCurrentApplication(null);
  }

  //useKeyboardEscape(handleCancel, isFormOpen);

  const renderCurrentView = () => {
    switch (currentView) {
      case 'timeline':
        return (
          <TimelineView
            applications={applications}
            onEdit={handleEdit}
            onDelete={handleDeleteEntry}
          />
        );
      case 'table':
      default:
        return (
          <ApplicationTable
            columns={initialColumns}
            data={applications}
            onEdit={handleEdit}
            onDelete={handleDeleteEntry}
          />
        );
      case 'kanban':
      case 'calendar':
        return (
          <div className="text-center py-20 text-gray-400">
            <p className="text-lg font-medium">Coming Soon!</p>
            <p className="text-sm mt-2">{currentView} view is under development</p>
          </div>
        );
    }
  };

  return (
    <AlertProvider>
      <div className="min-h-screen bg-gray-50">
        <Header />
        <main className="max-w-7xl mx-auto py-8 px-4 sm:px-6 lg:px-8">
          
          {/* Summary Section */}
          <MetricsSummary applications={applications} />
          
          {/* View Switcher, Header and Add Button */}
          <div className="flex flex-col sm:flex-row justify-between items-start sm:items-center gap-4 mb-6 mt-10">
            <div className="flex items-center gap-4 w-full sm:w-auto">
              <h2 className="text-2xl font-bold text-gray-800">Application Pipeline</h2>
              <ViewSwitcher currentView={currentView} onViewChange={setCurrentView} />
            </div>
            <button 
              className="bg-green-600 hover:bg-green-700 text-white font-bold py-2.5 px-6 rounded-full shadow-lg transition duration-150 transform hover:scale-[1.05]"
              onClick={handleCreateNew}
              aria-label="Add new application entry"
              data-testid="add-entry-button"
            >
              + Add Entry
            </button>
          </div>
          
          {/* Current View */}
          {renderCurrentView()}
        </main>
        {isFormOpen && (
          <AddJobForm 
            initialData={currentApplication} // Pasar datos para prellenar
            onSave={handleSaveEntry}
            onCancel={handleCancel}
          />
        )}
      </div>
    </AlertProvider>
  );
};

export default HomePage;