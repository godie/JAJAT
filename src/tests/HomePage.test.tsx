// src/tests/HomePage.test.tsx
import { render, screen, fireEvent, waitFor } from '@testing-library/react';
import HomePage from '../pages/HomePage';
import { expect, test, describe, beforeEach, vi } from 'vitest';

// 💡 1. Configuración de Mock para localStorage
// Esta es la implementación crucial para TDD con persistencia
const localStorageMock = (() => {
  let store: Record<string, string> = {};
  return {
    getItem: vi.fn((key: string) => store[key] || null),
    setItem: vi.fn((key: string, value: string) => {
      store[key] = value;
    }),
    clear: vi.fn(() => {
      store = {};
    }),
    removeItem: vi.fn((key: string) => {
      delete store[key];
    }),
    // Acceso para inspección en los tests
    getStore: () => store, 
  };
})();

// Reemplazar global.localStorage con nuestro mock
Object.defineProperty(global, 'localStorage', { value: localStorageMock });

// Requisito 6.4: Snapshot test for the main page.
test('HomePage renders correctly and matches snapshot', () => {
  const { container } = render(<HomePage />);
  // Un snapshot es vital para asegurar que los cambios de diseño no rompan la estructura.
  expect(container).toMatchSnapshot();
});

describe('HomePage Core Requirements', () => {
  
  beforeEach(() => {
    // Renderizar antes de cada prueba para limpieza
    render(<HomePage />);
  });

  // Requisito 6.1: Rendering the header and title.
  test('renders the application header and the correct title', () => {
    // Usar getByRole y getByTestId (del Header.tsx) para verificar el título
    expect(screen.getByText(/Job Application Tracker/i)).toBeInTheDocument();
  });
  
  test('renders the "Login with Google" button in the header', () => {
    // Verificar el botón de la cabecera
    const loginButton = screen.getByRole('button', { name: /Login with Google/i });
    expect(loginButton).toBeInTheDocument();
  });

  // Requisito 6.3: Ensuring the "Add Entry" button is visible.
  test('renders the "+ Add Entry" button for adding new applications', () => {
    const addButton = screen.getByRole('button', { name: "Add new application entry" });
    expect(addButton).toBeInTheDocument();
    expect(addButton).toBeVisible();
  });

  // Requisito 6.2: Rendering the table with correct columns.
  test('renders the application table with all required columns', () => {
    const requiredColumns = [
      'Position',
      'Company',
      'Salary',
      'Status',
      'Application Date',
      'Interview Date',
      'Platform',
      'Contact Name',
      'Follow-up Date',
      'Notes',
      'Link',
    ];

    // Verificar que todos los encabezados de columna estén presentes
    requiredColumns.forEach(column => {
      // Buscar por el rol 'columnheader' con el texto de la columna
      expect(screen.getByRole('columnheader', { name: column })).toBeInTheDocument();
    });
  });

  test('renders the summary metrics placeholders', () => {
    // Verificar que se rendericen los 3 placeholders de métricas
    const applicationMetrics = screen.getAllByText(/Applications/i);
    expect(applicationMetrics.length).toBeGreaterThanOrEqual(1);
    expect(screen.getByText(/Interviews/i)).toBeInTheDocument();
    expect(screen.getByText(/Offers/i)).toBeInTheDocument();
  });
});

describe('HomePage Core Functionality and Persistence', () => {
  
  beforeEach(() => {
    // Resetear el store y el componente antes de cada prueba
    localStorageMock.clear();
    localStorageMock.setItem('isLoggedIn', 'false'); // Estado inicial de login
    render(<HomePage />);
  });

  // Requisito 7.1: Rendering the header and login button
  test('renders the header, title, and Login button', () => {
    expect(screen.getByText(/Job Application Tracker/i)).toBeInTheDocument();
    
    // Verificar que el botón de Login/Logout esté visible
    const loginButton = screen.getByTestId('login-button');
    expect(loginButton).toBeInTheDocument();
    expect(loginButton).toHaveTextContent('Login with Google'); // Estado inicial
  });

  // Requisito 7.2: Rendering the table with correct columns
  test('renders the application table with all 11 required columns', () => {
    const requiredColumns = [
      'Position', 'Company', 'Salary', 'Status', 'Application Date', 
      'Interview Date', 'Platform', 'Contact Name', 'Follow-up Date', 'Notes', 'Link'
    ];
    requiredColumns.forEach(column => {
      expect(screen.getByRole('columnheader', { name: column })).toBeInTheDocument();
    });
  });

  // Requisito 7.3: Adding a new job entry updates localStorage
  test('clicking "+ Add Entry" button adds a new entry and persists it to localStorage', async () => {
    //const { rerender } = render(<HomePage />);
    localStorageMock.setItem.mockClear();
    const addButton = screen.getByTestId('add-entry-button');
    
    // 1. Verificar el estado inicial: localStorage debe estar vacío
    expect(localStorageMock.getItem('jobTrackerData')).toBeNull();

    // 2. Simular la adición de una entrada
    fireEvent.click(addButton);

    // 3. Esperar que React actualice el estado del componente (la tabla)
    await waitFor(() => {
        expect(screen.getByText(/New Position 1/i)).toBeInTheDocument();
    });

    // 4. Verificar el localStorage mock: debe tener 1 elemento
    const savedData = JSON.parse(localStorageMock.getItem('jobTrackerData')!);
    
    expect(localStorageMock.setItem).toHaveBeenCalledTimes(1); // La función saveApplications fue llamada
    expect(savedData).toHaveLength(1);
    expect(savedData[0].company).toBe('Placeholder Co.');
    expect(savedData[0].platform).toBe('LinkedIn'); 
    expect(savedData[0].contactName).toBe('N/A');

    // 5. Simular la adición de una segunda entrada
    fireEvent.click(addButton);
    
    // 6. Verificar que el array se actualizó a 2 elementos
    const updatedData = JSON.parse(localStorageMock.getItem('jobTrackerData')!);
    expect(updatedData).toHaveLength(2);
  });
});

// Requisito 7.4: Snapshot test for the main page.
// 💡 NOTA: Debes ejecutar 'npm test -- -u' para actualizar el snapshot
test('HomePage renders correctly and matches snapshot', () => {
  const { container } = render(<HomePage />);
  expect(container).toMatchSnapshot();
});