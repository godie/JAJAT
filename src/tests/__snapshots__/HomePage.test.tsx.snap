// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`HomePage renders correctly and matches snapshot 1`] = `
<div>
  <div
    class="min-h-screen bg-gray-50"
  >
    <header
      class="flex justify-between items-center p-4 border-b border-gray-200 bg-white shadow-sm"
    >
      <h1
        class="text-3xl font-extrabold text-indigo-700"
        data-testid="app-title"
      >
        Just Another Job Application Tracker
      </h1>
      <button
        aria-label="Login with Google"
        class="font-medium py-2 px-4 rounded-lg shadow-md transition duration-150 transform hover:scale-[1.02] bg-blue-600 hover:bg-blue-700 text-white "
        data-testid="login-button"
      >
        Login with Google
      </button>
    </header>
    <main
      class="max-w-7xl mx-auto py-8 px-4 sm:px-6 lg:px-8"
    >
      <section
        class="grid grid-cols-1 sm:grid-cols-2 lg:grid-cols-3 gap-6 my-8"
        data-testid="metrics-summary"
      >
        <div
          class="bg-white p-6 rounded-xl shadow-lg border-l-4 border-blue-500 transition duration-300 hover:shadow-xl"
        >
          <p
            class="text-sm font-medium text-gray-500"
          >
            Applications
          </p>
          <p
            class="mt-1 text-4xl font-extrabold text-gray-900"
          >
            0
          </p>
        </div>
        <div
          class="bg-white p-6 rounded-xl shadow-lg border-l-4 border-yellow-500 transition duration-300 hover:shadow-xl"
        >
          <p
            class="text-sm font-medium text-gray-500"
          >
            Interviews
          </p>
          <p
            class="mt-1 text-4xl font-extrabold text-gray-900"
          >
            0
          </p>
        </div>
        <div
          class="bg-white p-6 rounded-xl shadow-lg border-l-4 border-green-500 transition duration-300 hover:shadow-xl"
        >
          <p
            class="text-sm font-medium text-gray-500"
          >
            Offers
          </p>
          <p
            class="mt-1 text-4xl font-extrabold text-gray-900"
          >
            0
          </p>
        </div>
      </section>
      <div
        class="flex flex-col sm:flex-row justify-between items-start sm:items-center gap-4 mb-6 mt-10"
      >
        <div
          class="flex items-center gap-4 w-full sm:w-auto"
        >
          <h2
            class="text-2xl font-bold text-gray-800"
          >
            Application Pipeline
          </h2>
          <div
            class="flex items-center space-x-2 bg-white rounded-lg shadow-sm border border-gray-200 p-2"
          >
            <button
              class="
              flex items-center space-x-2 px-4 py-2 rounded-lg transition-all duration-150
              bg-indigo-100 text-indigo-700 font-semibold shadow-sm
            "
              title="Enhanced table with filters"
            >
              <svg
                class="w-5 h-5"
                fill="none"
                stroke="currentColor"
                viewBox="0 0 24 24"
              >
                <path
                  d="M3 10h18M3 14h18m-9-4v8m-7 0h14a2 2 0 002-2V8a2 2 0 00-2-2H5a2 2 0 00-2 2v8a2 2 0 002 2z"
                  stroke-linecap="round"
                  stroke-linejoin="round"
                  stroke-width="2"
                />
              </svg>
              <span
                class="hidden sm:inline"
              >
                Table
              </span>
            </button>
            <button
              class="
              flex items-center space-x-2 px-4 py-2 rounded-lg transition-all duration-150
              text-gray-600 hover:bg-gray-100 hover:text-gray-900
            "
              title="Chronological interview flow"
            >
              <svg
                class="w-5 h-5"
                fill="none"
                stroke="currentColor"
                viewBox="0 0 24 24"
              >
                <path
                  d="M12 8v4l3 3m6-3a9 9 0 11-18 0 9 9 0 0118 0z"
                  stroke-linecap="round"
                  stroke-linejoin="round"
                  stroke-width="2"
                />
              </svg>
              <span
                class="hidden sm:inline"
              >
                Timeline
              </span>
            </button>
            <button
              class="
              flex items-center space-x-2 px-4 py-2 rounded-lg transition-all duration-150
              text-gray-400 cursor-not-allowed opacity-50
            "
              disabled=""
              title="Board view [Coming Soon]"
            >
              <svg
                class="w-5 h-5"
                fill="none"
                stroke="currentColor"
                viewBox="0 0 24 24"
              >
                <path
                  d="M9 17V7m0 10a2 2 0 01-2 2H5a2 2 0 01-2-2V7a2 2 0 012-2h2a2 2 0 012 2m0 10a2 2 0 002 2h2a2 2 0 002-2M9 7a2 2 0 012-2h2a2 2 0 012 2m0 10V7m0 10a2 2 0 002 2h2a2 2 0 002-2V7a2 2 0 00-2-2h-2a2 2 0 00-2 2"
                  stroke-linecap="round"
                  stroke-linejoin="round"
                  stroke-width="2"
                />
              </svg>
              <span
                class="hidden sm:inline"
              >
                Kanban
              </span>
            </button>
            <button
              class="
              flex items-center space-x-2 px-4 py-2 rounded-lg transition-all duration-150
              text-gray-400 cursor-not-allowed opacity-50
            "
              disabled=""
              title="Calendar view [Coming Soon]"
            >
              <svg
                class="w-5 h-5"
                fill="none"
                stroke="currentColor"
                viewBox="0 0 24 24"
              >
                <path
                  d="M8 7V3m8 4V3m-9 8h10M5 21h14a2 2 0 002-2V7a2 2 0 00-2-2H5a2 2 0 00-2 2v12a2 2 0 002 2z"
                  stroke-linecap="round"
                  stroke-linejoin="round"
                  stroke-width="2"
                />
              </svg>
              <span
                class="hidden sm:inline"
              >
                Calendar
              </span>
            </button>
          </div>
        </div>
        <button
          aria-label="Add new application entry"
          class="bg-green-600 hover:bg-green-700 text-white font-bold py-2.5 px-6 rounded-full shadow-lg transition duration-150 transform hover:scale-[1.05]"
          data-testid="add-entry-button"
        >
          + Add Entry
        </button>
      </div>
      <div
        class="overflow-x-auto shadow-xl rounded-lg border border-gray-100"
      >
        <table
          class="min-w-full divide-y divide-gray-200"
          data-testid="application-table"
        >
          <thead
            class="bg-gray-50"
          >
            <tr>
              <th
                class="px-6 py-3 text-left text-xs font-semibold text-gray-600 uppercase tracking-wider bg-indigo-50 whitespace-nowrap"
                scope="col"
              >
                Position
              </th>
              <th
                class="px-6 py-3 text-left text-xs font-semibold text-gray-600 uppercase tracking-wider bg-indigo-50 whitespace-nowrap"
                scope="col"
              >
                Company
              </th>
              <th
                class="px-6 py-3 text-left text-xs font-semibold text-gray-600 uppercase tracking-wider bg-indigo-50 whitespace-nowrap"
                scope="col"
              >
                Salary
              </th>
              <th
                class="px-6 py-3 text-left text-xs font-semibold text-gray-600 uppercase tracking-wider bg-indigo-50 whitespace-nowrap"
                scope="col"
              >
                Status
              </th>
              <th
                class="px-6 py-3 text-left text-xs font-semibold text-gray-600 uppercase tracking-wider bg-indigo-50 whitespace-nowrap"
                scope="col"
              >
                Application Date
              </th>
              <th
                class="px-6 py-3 text-left text-xs font-semibold text-gray-600 uppercase tracking-wider bg-indigo-50 whitespace-nowrap"
                scope="col"
              >
                Interview Date
              </th>
              <th
                class="px-6 py-3 text-left text-xs font-semibold text-gray-600 uppercase tracking-wider bg-indigo-50 whitespace-nowrap"
                scope="col"
              >
                Platform
              </th>
              <th
                class="px-6 py-3 text-left text-xs font-semibold text-gray-600 uppercase tracking-wider bg-indigo-50 whitespace-nowrap"
                scope="col"
              >
                Contact Name
              </th>
              <th
                class="px-6 py-3 text-left text-xs font-semibold text-gray-600 uppercase tracking-wider bg-indigo-50 whitespace-nowrap"
                scope="col"
              >
                Follow-up Date
              </th>
              <th
                class="px-6 py-3 text-left text-xs font-semibold text-gray-600 uppercase tracking-wider bg-indigo-50 whitespace-nowrap"
                scope="col"
              >
                Notes
              </th>
              <th
                class="px-6 py-3 text-left text-xs font-semibold text-gray-600 uppercase tracking-wider bg-indigo-50 whitespace-nowrap"
                scope="col"
              >
                Link
              </th>
            </tr>
          </thead>
          <tbody
            class="bg-white divide-y divide-gray-100"
          >
            <tr>
              <td
                class="px-6 py-10 text-center text-gray-400 italic"
                colspan="12"
              >
                Use the "+ Add Entry" button to start tracking your applications!
              </td>
            </tr>
          </tbody>
        </table>
      </div>
    </main>
  </div>
  <div
    class="fixed top-4 right-4 z-50 w-full max-w-md space-y-2 pointer-events-none"
  />
</div>
`;
